arc

```diff
@@ -0,0 +1,163 @@
+
+shift "$((OPTIND - 1))"
+
+debug "Parsed options"
+
+while [[ $# -gt 0 ]]; do
+    case $1 in
+        --)
+            shift
+            break
+            ;;
+        check)
+            MODE=check
+            INPUT_FILE="$1"
+            shift 3
+            break
+            ;;
+        build)
+            MODE=build
+            INPUT_FILE="$2"
+            shift 3
+            break
+            ;;
+        run)
+            MODE=run
+            INPUT_FILE="$2"
+            shift 3
+            break
+            ;;
+        *)
+            echo "Invalid subcommand: $1" >&2
+            usage
+            exit 1
+            ;;
+    esac
+done
+
+debug "Parsed subcommand"
+
+CRATE_NAME="$(basename $INPUT_FILE .arc)-crate"
+
+if [ -z "$INPUT_FILE" ]; then
+    usage
+fi
+
+if [ -z "$CRATE_DIR" ]; then
+    CRATE_DIR="$(dirname $INPUT_FILE)/$CRATE_NAME"
+fi
+
+ARC_DIR=$(realpath $(dirname "$(which "$0")")/)
+
+if [ -z "$ARC_LANG" ]; then
+    ARC_LANG="$ARC_DIR/arc-mlir/build/llvm-build/bin/arc-lang"
+    if [ ! -f "$ARC_LANG" ]; then
+        echo "Could not find arc-lang. Did you run './build'?"
+        exit 1
+    fi
+fi
+
+if [ -z "$ARC_MLIR" ]; then
+    ARC_MLIR="$ARC_DIR/arc-mlir/build/llvm-build/bin/arc-mlir"
+    if [ ! -f "$ARC_LANG" ]; then
+        echo "Could not find arc-mlir. Did you run './build'?"
+        exit 1
+    fi
+fi
+
+if [ -z "$ARC_RUNTIME_DIR" ]; then
+    ARC_RUNTIME_DIR="$ARC_DIR/arc-runtime/"
+fi
+
+INPUT_FILE="$(realpath $INPUT_FILE)"
+CRATE_DIR="$(realpath $CRATE_DIR)"
+CRATE_MAIN_FILE="$CRATE_DIR/src/main.rs"
+CRATE_TOML_FILE="$CRATE_DIR/Cargo.toml"
+
+debug "INPUT_FILE: $INPUT_FILE"
+debug "CRATE_DIR: $CRATE_DIR"
+debug "CRATE_MAIN_FILE: $CRATE_MAIN_FILE"
+debug "CRATE_TOML_FILE: $CRATE_TOML_FILE"
+debug "CRATE_NAME: $CRATE_NAME"
+debug "ARC_LANG: $ARC_LANG"
+debug "ARC_MLIR: $ARC_MLIR"
+debug "ARC_DIR: $ARC_MLIR"
+debug "ARC_RUNTIME_DIR: $ARC_RUNTIME_DIR"
+debug "MLIR args: $@"
+
+mkdir -p "$CRATE_DIR/src"
+
+# Create the Cargo.toml
+envsubst > "$CRATE_DIR/Cargo.toml" <<'EOF'
+[package]
+name        = "$CRATE_NAME"
+version     = "0.0.0"
+edition     = "2021"
+license     = "MIT"
+description = "Generic MLIR-to-Rust test case"
+
+[dependencies]
+arc-runtime = { version = "=0.0.0", path = "$ARC_RUNTIME_DIR", features = ["legacy"] }
+hexf        = { version = "0.2.1" }
+ndarray     = { version = "0.13.0" }
+prost       = { version = "0.7.0" }
+
+[profile.dev]
+debug = false
+EOF
+
+debug "Cargo.toml: $(cat $CRATE_DIR/Cargo.toml)"
+
+echo '#![feature(unboxed_closures)]' > "$CRATE_MAIN_FILE"
+$ARC_LANG "$INPUT_FILE" | $ARC_MLIR "$@" -arc-to-rust -inline-rust -arc-lang-runtime >> "$CRATE_MAIN_FILE"
+
+case "$MODE" in
+    check)
+        cargo check --manifest-path "$CRATE_TOML_FILE"
+        ;;
+    build)
+        cargo build --manifest-path "$CRATE_TOML_FILE"
+        ;;
+    run)
+        cargo run --manifest-path "$CRATE_TOML_FILE"
+        ;;
+esac
+
+debug "Generated code: $(cat $CRATE_MAIN_FILE)"
```

Comment:
This could probably overflow the maximum size of a command line argument.

You could do:
```
 [ "$DEBUG" ] && echo "Generated code" && cat $"CRATE_MAIN_FILE"
```